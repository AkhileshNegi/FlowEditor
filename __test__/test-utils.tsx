import { ReactWrapper } from 'enzyme';
import { ReactElement } from 'react';
import { FlowDefinition } from '../src/FlowDefinition';

const request = require('sync-request');

function getFlow(name: string): FlowDefinition {
    const definition = request('GET', 'base/test_flows/' + name + '.json').getBody();
    return JSON.parse(definition) as FlowDefinition;
}

export function getFavorites(): FlowDefinition {
    return getFlow('favorites');
}

export function getTest(): FlowDefinition {
    return getFlow('a4f64f1b-85bc-477e-b706-de313a022979');
}

export function dump(object: any) {
    console.log(JSON.stringify(object, null, 1));
}

export const DATA_SPEC_ATTRIBUTE_NAME: string = 'data-spec';

/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param - Name of `data-spec` attribute value to find
 * @param typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns All matching DOM components
 */
export const getSpecWrapper = (
    componentWrapper: ReactWrapper,
    specName: string,
    typeFilter: string
): ReactElement<any> => {
    let specWrappers;

    if (!typeFilter) {
        specWrappers = componentWrapper.find(`[${DATA_SPEC_ATTRIBUTE_NAME}="${specName}"]`);
    } else {
        specWrappers = componentWrapper.findWhere(
            wrapper =>
                wrapper.prop(DATA_SPEC_ATTRIBUTE_NAME) === specName && wrapper.type() === typeFilter
        );
    }

    return specWrappers;
};
