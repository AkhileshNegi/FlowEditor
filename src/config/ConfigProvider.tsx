import * as React from 'react';
import { Endpoints, FlowEditorConfig } from '~/flowTypes';

export const fakePropType: any = (): any => null;
fakePropType.isRequired = (): any => null;

interface ConfigProviderProps {
    config: FlowEditorConfig;
    children: any;
}

export interface ConfigProviderContext {
    endpoints: Endpoints;
    flow: string;
    debug: boolean;
    showDownload: boolean;
}

// ----------------------------------------------------------------------------------------------

export const SINGLE_CHILD_ERROR = 'ConfigProvider expects only one child component.';
export const VALID_CHILD_ERROR =
    'ConfigProvider expects a valid React element: https://reactjs.org/docs/react-api.html#isvalidelement';

export default class ConfigProvider extends React.Component<ConfigProviderProps> {
    public static childContextTypes = {
        assetService: fakePropType,
        endpoints: fakePropType,
        flow: fakePropType,
        debug: fakePropType,
        showDownload: fakePropType
    };

    constructor(props: ConfigProviderProps) {
        super(props);

        if (React.Children.count(props.children) > 1) {
            throw new Error(SINGLE_CHILD_ERROR);
        } else if (!React.isValidElement(props.children)) {
            throw new Error(VALID_CHILD_ERROR);
        }
    }

    public getChildContext(): ConfigProviderContext {
        return {
            endpoints: this.props.config.endpoints,
            flow: this.props.config.flow,
            debug: this.props.config.debug,
            showDownload: this.props.config.showDownload
        };
    }

    public render(): JSX.Element {
        return React.Children.only(this.props.children);
    }
}
