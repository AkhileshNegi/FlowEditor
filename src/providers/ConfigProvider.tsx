import * as React from 'react';
import { Endpoints, Languages } from '../flowTypes';
import { Language } from '../components/LanguageSelector';
import { flow } from 'Config';
import {
    Type,
    TypeMap,
    GetTypeConfig,
    typeConfigList,
    actionConfigList,
    typeConfigMap,
    getTypeConfig
} from './typeConfigs';
import {
    Operator,
    OperatorMap,
    GetOperatorConfig,
    operatorConfigList,
    operatorConfigMap,
    getOperatorConfig
} from './operatorConfigs';
import {
    GetActivity,
    GetFlows,
    GetFlow,
    SaveFlow,
    getActivity,
    getFlows,
    getFlow,
    saveFlow
} from './external';
import { childContextPT } from './propTypes';

export const getBaseLanguage = (languages: Languages): Language => {
    const [iso] = Object.keys(languages);
    const name = languages[iso];
    return {
        name,
        iso
    };
};

export interface ConfigProviderProps {
    flowEditorConfig: FlowEditorConfig;
}

export interface ConfigProviderContext {
    typeConfigList: Type[];
    operatorConfigList: Operator[];
    actionConfigList: Type[];
    typeConfigMap: TypeMap;
    operatorConfigMap: OperatorMap;
    getTypeConfig: GetTypeConfig;
    getOperatorConfig: GetOperatorConfig;
    getActivity: GetActivity;
    getFlows: GetFlows;
    getFlow: GetFlow;
    saveFlow: SaveFlow;
    flow: string;
    baseLanguage: Language;
    endpoints: Endpoints;
    languages: Languages;
}

export default class ConfigProvider extends React.Component<ConfigProviderProps> {
    public static childContextTypes = childContextPT;

    constructor(props: ConfigProviderProps) {
        super(props);
    }

    public getChildContext(): ConfigProviderContext {
        const { flowEditorConfig: { endpoints, languages } } = this.props;
        const baseLanguage = getBaseLanguage(languages);
        return {
            typeConfigList,
            operatorConfigList,
            actionConfigList,
            typeConfigMap,
            operatorConfigMap,
            getTypeConfig,
            getOperatorConfig,
            getActivity,
            getFlows,
            getFlow,
            saveFlow,
            flow,
            baseLanguage,
            endpoints,
            languages
        };
    }

    public render(): JSX.Element {
        return React.Children.only(this.props.children);
    }
}
