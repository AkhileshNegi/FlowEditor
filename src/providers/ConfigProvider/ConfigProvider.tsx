import * as React from 'react';
import { ConfigProviderContext } from './configContext';
import { childContextPT } from './propTypes';
import configContext from './configContext';

export interface ConfigProviderProps {
    flowEditorConfig: FlowEditorConfig;
}


export default class ConfigProvider extends React.Component<ConfigProviderProps> {
    public static childContextTypes = childContextPT;

    constructor(props: ConfigProviderProps) {
        super(props);
        if (React.Children.count(this.props.children) > 1 ) {
            throw new Error('ConfigProvider expects to receive only one child component.');
        }
    }

    public getChildContext(): ConfigProviderContext {
        return configContext;
    }

    public render(): JSX.Element {
        return React.Children.only(this.props.children);
    }
}
