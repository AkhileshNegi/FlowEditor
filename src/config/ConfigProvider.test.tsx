import { shallow } from 'enzyme';
import * as React from 'react';
import { findRenderedComponentWithType, renderIntoDocument } from 'react-dom/test-utils';

import ConfigProvider, {
    fakePropType,
    SINGLE_CHILD_ERROR,
    VALID_CHILD_ERROR
} from './ConfigProvider';

const config = require('../../__test__/config.js');

describe('ConfigProvider >', () => {
    const createChild = () => {
        class Child extends React.Component<{}> {
            public static contextTypes = {
                endpoints: fakePropType,
                languages: fakePropType,
                flow: fakePropType
            };
            public render(): JSX.Element {
                return <div />;
            }
        }
        return Child;
    };

    it('should throw if more than one child component provided', () => {
        expect(() =>
            shallow(
                <ConfigProvider config={config}>
                    <div />
                    <div />
                </ConfigProvider>
            )
        ).toThrowError(SINGLE_CHILD_ERROR);
    });

    it('should throw if child is not a valid React element', () => {
        expect(() =>
            shallow(
                <ConfigProvider config={config}>
                    {() => {
                        return;
                    }}
                </ConfigProvider>
            )
        ).toThrowError(VALID_CHILD_ERROR);
    });

    it("should provide config to child's context", () => {
        const Child = createChild();
        const tree = renderIntoDocument(
            <ConfigProvider config={config}>
                <Child />
            </ConfigProvider>
        ) as React.Component<any, {}>;
        const childComp = findRenderedComponentWithType(tree, Child);
        const { assetHost, endpoints, languages, flow } = config;

        expect(childComp.context).toEqual({
            endpoints,
            languages,
            flow
        });
        expect(childComp.context).toMatchSnapshot();
    });
});
