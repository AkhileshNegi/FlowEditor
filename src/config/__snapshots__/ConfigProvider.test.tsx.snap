// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider > should provide config to child's context 1`] = `
Object {
  "endpoints": Object {
    "activity": "",
    "fields": "/assets/fields.json",
    "flows": "/assets/flows.json",
    "groups": "/assets/groups.json",
    "recipients": "/assets/recipients.json",
    "simulateResume": "",
    "simulateStart": "",
  },
  "flow": "a4f64f1b-85bc-477e-b706-de313a022979",
  "languages": Object {
    "eng": "English",
    "fre": "French",
    "spa": "Spanish",
  },
}
`;
